### PR TITLE
Initial parser implementation returning generic enum instead of groups

### PR DESCRIPTION
**Pull request by Artemkaaas**
This option reflects parside implementation which return generic enum as result of the work (without particular groups definition). 
In this case we need to create a third crate combining parside and cerside  and defining groups constructors + helpers. This crate will produce single library which will be consuming by other applications. 

An alternative approach to: https://github.com/WebOfTrust/parside/pull/2

### CODE REVIEW

**Review by jasoncolburne — COMMENTED** [review]
I think I prefer this implementation for these reasons:
- less maintenance overhead
- less to reason about
- cleaner in general
Drawbacks:
- these higher level concepts may be harder for newcomers to grok (but once this is built, is that a concern?)
- need another library.

But, to make an educated decision I would need an estimate of the scope (size and complexity) of the third library - you are better suited than I to understand what is required there. It would incur additional overhead for yet another library, and if it's going to be roughly the same scope as #2 (which approaches 4k lines) it's a much harder decision to make. @m00sey what do you think?

Can you or Dmitry provide further insight?

**Comment by Artemkaaas** [comment]
```
Can you (Artem) or Dmitry provide further insight on the scope of the required lib?
```
I've already prepared [initial implementation for it as well](https://github.com/DSRCorporation/cesrox/tree/integrate-with-cesride)

**Comment by jasoncolburne** [comment]
The `cesrox` branch actually looks cleaner to me than #2. It's only about 1000 lines. I checked it out and did a local diff on `src/message` to figure this out. I am still not sure how I feel about another library, and being opinionated about how consumers of this code consume.

File: Cargo.toml [hunk 1]
```diff
@@ -8,4 +8,12 @@ license = "Apache-2.0"
...
 
 [dependencies]
-cesride = "0.1.3"
+cesride = { git = "https://github.com/DSRCorporation/cesride.git" }
+nom = "5"
+num-derive = "0.3.3"
+num-traits = "0.2"
+rmp-serde = "0.15"
+serde = { version = "1.0", features = ["derive"] }
+serde_json = "1.0"
```
Comment (jasoncolburne) [inline Cargo.toml:17]:
```suggestion
serde_json = { version = "1.0", features = ["preserve_order"] }
```

In this context I believe this is a requirement.

Comment (jasoncolburne) [reply Cargo.toml:17]: https://docs.rs/crate/serde_json/1.0.93/features

**Comment by jasoncolburne** [comment]
I'm going to check out these branches over the weekend and actually play around with them. I think we should optimize for consumers, and this is the easiest way for me to evaluate.

**Comment by Artemkaaas** [comment]
It looks like this approach will not work anymore with the latest `cesride` changes (`Matter` became a `trait` and other classes like `Verfer` implement this trait ). 
The issue is that we cannot return a generic list `type Single = Vec<Box<dyn Matter>>;` because `Matter` trait contains `static` methods. And we cannot correctly create a sized object doing `Matter::new_with_qb64b`.
Now we must define a list of exact structures (`type Single = Vec<Verfer>;`) and create them. in fact looks like we are forced to use the option defined in [PR #2](https://github.com/WebOfTrust/parside/pull/3).

**Comment by jasoncolburne** [comment]
So let me see, there is trait with static methods and also abstract methods? What about .... TWO TRAITS? One that is public that exposes the callbacks, and one that depends on that and implements the static functionality?

**Comment by Artemkaaas** [comment]
I may misunderstand your idea.  
1. Parside uses static functions like `new_with_qb64b` to create instances.   
2. Now Parside have to call these functions `new_with_qb64b` with specific type like `Dater:new_with_qb64b`.
3. It means that Parside needs to know the exact definition of a group. 

For instance, each item of `Codex::NonTransReceiptCouples` and `Codex::FirstSeenReplayCouples` groups consists of two different primitives `(Verfer+Cigar) and (Seqner+Dater)` . So in order to parse these groups properly Parside need to call the corresponding methods.

Am I wrong?

**Comment by jasoncolburne** [comment]
I refactored all those methods into single `new` interface so it would better match the python implementation in cesride#93. It should make the implementation easier to get through `uniffi`, based on my experience. And it reduced code quite a bit.

What I've seen that may solve this problem is an Enum. You can create a variant for each type but refer to the enum itself when accepting inputs, allowing you to use any of the types with an enum implementation that delegates in a match statement. The enum would need a single wrapper for each of the exposed methods, and it would match on type. In most matches, you would return an error. But a few for each implementation would instantiate the class. Does this solve your problem? I can write it out if you want

**Comment by jasoncolburne** [comment]
Or maybe I still misunderstand. I'll look more closely at the code here

**Comment by jasoncolburne** [comment]
I think I already understand the problem with my proposed approach.. We can't return polymorphic types in Rust right? It would be simple if we could but that doesn't make sense.

**Comment by Artemkaaas** [comment]
Do you mean having enum like this?
```
enum CesrPrimitive {
   Verfer(val),
   Dater(val).
   Prefixes(val),
   Seqner(val),
   ....   
}
```
Do all primitives parses uniquely or depending on a condition/group?
I mean if inside of the `parse_primitive` function we will apply the parser functions `Verfer::new, Dater::new, ....` one by one until the first success, will we receive the correct result always?

**Comment by jasoncolburne** [comment]
You wouldn't need to go one by one, you could match against self, which would tell you the type. But then, to get the output (the new object) back out of the function, you need to type that somehow, which you'd need to define a trait or set of traits for I believe.

**Pull request closed without merging**